pkg/apis/portworx/v1beta2/types.go

```diff
@@ -0,0 +1,75 @@
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:noStatus
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumePlacementStrategy specifies a spec for volume placement in the cluster
+type VolumePlacementStrategy struct {
+	meta.TypeMeta   `json:",inline"`
+	meta.ObjectMeta `json:"metadata,omitempty"`
+	Spec            VolumePlacementSpec `json:"spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumePlacementStrategyList is a list of VolumePlacementStrategy objects
+type VolumePlacementStrategyList struct {
+	meta.TypeMeta `json:",inline"`
+	meta.ListMeta `json:"metadata,omitempty"`
+	// Items are the list of volume placements strategy items
+	Items []VolumePlacementStrategy `json:"items"`
+}
+
+// VolumePlacementSpec specifies a set of rules for volume placement in the cluster
+type VolumePlacementSpec struct {
+	// Rules defines a list of rules as part of the placement spec. All the rules specified will
+	// be applied for volume placement.
+	// Rules that have enforcement as "required" are strictly enforced while "preferred" are best effort.
+	// In situations, where 2 or more rules conflict, the weight of the rules will dictate which wins.
+	Rules VolumePlacementRule `json:"rules"`
+}
+
+// VolumePlacementRule defines the rule for placing volume replicas
+type VolumePlacementRule struct {
+	// ReplicaAffinity defines affinity rules between replicas within a volume
+	ReplicaAffinity []*ReplicaPlacementSpec `json:"replicaAffinity,omitempty"`
+	// ReplicaAntiAffinity defines anti-affinity rules between replicas within a volume
+	ReplicaAntiAffinity []*ReplicaPlacementSpec `json:"replicaAntiAffinity,omitempty"`
+	// VolumeAffinity defines affinity rules between volumes
+	VolumeAffinity []*CommonPlacementSpec `json:"volumeAffinity,omitempty"`
+	// VolumeAntiAffinity defines anti-affinity rules between volumes
+	VolumeAntiAffinity []*CommonPlacementSpec `json:"volumeAntiAffinity,omitempty"`
+}
+
+// ReplicaPlacementSpec is the spec for replica affinity and anti-affinity
+type ReplicaPlacementSpec struct {
+	CommonPlacementSpec
+	// AffectedReplicas defines the number of volume replicas affected by this rule. If not provided,
+	// rule would affect all replicas
+	// (optional)
+	AffectedReplicas int64 `json:"affected_replicas,omitempty"`
+}
+
+// CommonPlacementSpec is the spec that's common for replica and volume affinity and anti-affinity rules
+type CommonPlacementSpec struct {
+	// Weight defines the weight of the rule which allows to break the tie with other matching rules. A rule with
+	// higher weight wins over a rule with lower weight.
+	// (optional)
+	Weight int64 `json:"weight,omitempty"`
+	// Enforcement specifies the rule enforcement policy. Can take values: required or preferred.
+	// (optional)
+	Enforcement v1beta1.EnforcementType `json:"enforcement,omitempty"`
```

Comment:
Is it a good idea or known pattern to reference objects from older (v1beta1) versions?

Reply:
There is no known pattern to do this. I just didn't see any downside of doing this. Wanted to avoid duplicating the exact same type at 2 places.